extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/filters/CORSFilter.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.undertow.runtime.filters;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+public class CORSFilter implements Filter {
+
+    @ConfigProperty(name = "quarkus.http.cors.origins")
+    public Optional<String> origins;
+    @ConfigProperty(name = "quarkus.http.cors.methods")
+    public Optional<String> methods;
+    @ConfigProperty(name = "quarkus.http.cors.headers")
+    public Optional<String> headers;
+    @ConfigProperty(name = "quarkus.http.cors.exposed-headers")
+    public Optional<String> exposedHeaders;
```

Comment:
I don't like that. All the configuration should be in `CORSConfig`. Keep in mind that we plan to have an automatic configuration doc export. So having everything in proper config classes with javadoc is key.

I think you need to pass them as init params of the filter.

Reply:
I understand, I will do this.

Reply:
If you use ServletExtension you can use ImmediateInstanceFactory when you add the Filter to just pass a pre-constructed filter to Undertow. You could just make these final fields, read the config in the template and pass it directly to the constructor.

Reply:
@stuartwdouglas I am a little stuck here... I haven't found the way to access the HttpConfig in the ServletExtension as the extensions are loaded in a STATIC_INIT build step... I have done some deep debugging at the undertow extension and as I understand it is a two step process:

* First, build the servlet deployment with ALL the servlet extensions, web metadata, filters, servlets...etc (this happens at STATIC_INIT)
* Second,start Undertow with the deployment generated in step 1 (this happens at RUNTIME_INIT)

In this case the `handleDeployment` from my ServletExtension is called at STATIC_INIT and Quarkus is raising an error because I try to access the ConfigPhase.RUN_TIME object.

```
Caused by: java.lang.IllegalArgumentException: Bytecode recorder is static but an injected config object is declared as run time at public void io.quarkus.undertow.deployment.UndertowBuildStep.buildCorsFilter(io.quarkus.undertow.runtime.filters.CORSTemplate,io.quarkus.undertow.runtime.HttpBuildConfig,io.quarkus.undertow.runtime.HttpConfig,io.quarkus.deployment.annotations.BuildProducer) of class io.quarkus.undertow.deployment.UndertowBuildStep
	at io.quarkus.deployment.ExtensionLoader.reportError(ExtensionLoader.java:541)
	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:454)
	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:90)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:74)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:102)
	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:220)
```

The easy fix would be to access the config as a MP-Config, but I was trying to not duplicate the property names and access the original Quarkus config object. I wonder if it is possible ?

Thanks for your help.

Reply:
Sorry for the delayed reply, I have been travelling and missed the notification. 

I don't think there is a nice way to handle this at the moment, we don't really have a concept of config that is passed to the template at static init, but not actually used till runtime. 

The yuck but simple way to make this work is to have a static field on the template to store the config, and set it at runtime, then just have the filter delegate to this static field for its config.

A less yuck way is to create the CorsFilter in a template in a STATIC_INIT method, then create a RUNTIME_INIT method that takes the CorsFilter and the HttpConfig and explicitly sets the config on the filter.

We may need to look at a better way to handle this. In general I have just been trying to handle stuff like this on a case by case basis as it arrises, and if we see any common patterns/problems then try and abstract a common solution.